pkg/log/log_file.go

```diff
@@ -0,0 +1,171 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package main under directory cmd parses and validates user input,
+// instantiates and initializes objects imported from pkg, and runs
+// the process.
+package log
+
+import (
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/spf13/pflag"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+)
+
+const (
+	logToStdErrFlag = "logtostderr"
+	logDirFlag      = "log_dir"
+	logFileFlag     = "log_file"
+	maxSizeFlag     = "log_file_max_size"
+	maxNumFlag      = "log_file_max_num"
+
+	// Check INFO log file number every 10 mins.
+	logFileCheckInterval = time.Minute * 10
+)
+
+var (
+	logFileMaxNum = uint16(0)
+	logDir        = ""
+
+	executableName = filepath.Base(os.Args[0])
+)
+
+func AddFlags(fs *pflag.FlagSet) {
+	fs.Uint16Var(&logFileMaxNum, maxNumFlag, logFileMaxNum, "Maximum number of INFO log files to be kept. Value 0 means unlimited. Number of non-INFO log files is always unlimited")
+}
+
+// InitLogFileLimits initializes log file maximum size and maximum number limits based on the
+// command line flags.
+func InitLogFileLimits(fs *pflag.FlagSet) {
+	var err error
+	var logToStdErr bool
+	var logFile string
+	var maxSize uint64
+
+	logToStdErr, err = fs.GetBool(logToStdErrFlag)
+	if err != nil {
+		// Should not happen. Return for safety.
+		return
+	}
+	if logToStdErr {
+		// Not log to files.
+		return
+	}
+
+	logFile, err = fs.GetString(logFileFlag)
+	if err != nil {
+		return
+	}
+	if logFile != "" {
+		// Log to a single file. klog will take care of the max size limit.
+		return
+	}
+
+	maxSize, err = fs.GetUint64(maxSizeFlag)
+	if err != nil {
+		return
+	}
+
+	// klog does not respect the max file szie specified by --log_file_max_size
+	// when --log_file is not used. Here as a workaround, we directly set the
+	// the specified max size to klog.MaxSize.
+	if klog.MaxSize != maxSize {
+		klog.MaxSize = maxSize
+		klog.Infof("Set log file max size to %d", maxSize)
+	}
+
+	if logFileMaxNum > 0 {
+		logDir, err = fs.GetString(logDirFlag)
+		if err != nil {
+			// Should not happen. Disable the log file number limit for safety.
+			logFileMaxNum = 0
+		} else if logDir == "" {
+			// Log to the current dir.
+			logDir = os.TempDir()
+		}
+	}
+}
+
+// StartLogFileNumberMonitor starts monitoring the log files to make sure the
+// number of INFO log files does not exceed the maximum limit, when the log file
+// number limit is configured.
+func StartLogFileNumberMonitor(stopCh <-chan struct{}) {
+	if logFileMaxNum == 0 {
+		// The maximum log file number limit is not configured.
+		return
+	}
+
+	go func(stopCh <-chan struct{}) {
```

Comment:
do you need to have `stopCh` as an actual parameter here? Won't it be captured by go?

Reply:
Yes, it could be removed too.